////////////////////////////////////////////////////////////////////////////////////////////////////
///
/// @file      ProductMessage.proto
///
/// @brief     This header file declares a Protocol Buffer for creating messages to send as events
///            to the Product Controller state machine.
///
/// @author    Stuart J. Lumby
///
/// @date      09/22/2017
///
/// @attention Copyright (C) 2017 Bose Corporation All Rights Reserved
///
///            Bose Corporation
///            The Mountain Road,
///            Framingham, MA 01701-9168
///            U.S.A.
///
///            This program may not be reproduced, in whole or in part, in any form by any means
///            whatsoever without the written permission of Bose Corporation.
///
////////////////////////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////////////////////////
///                                STS Protocol Buffer Include                                   ///
////////////////////////////////////////////////////////////////////////////////////////////////////
import "ProductSTS.proto";

////////////////////////////////////////////////////////////////////////////////////////////////////
///                           System Language Protocol Buffer Include                            ///
////////////////////////////////////////////////////////////////////////////////////////////////////
import "Language.proto";

////////////////////////////////////////////////////////////////////////////////////////////////////
///                           Key Actions Protocol Buffer Include                                ///
////////////////////////////////////////////////////////////////////////////////////////////////////
import "KeyActions.proto";

////////////////////////////////////////////////////////////////////////////////////////////////////
///                       LPM Lower Power Microprocessor Message Structure                       ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductLpmStatus
{
        optional bool connected = 1;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                    CAPS Content Audio Playback Services Message Structure                    ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductCapsStatus
{
        optional bool initialized = 1;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                            Audio Path Status Message Structure                               ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductAudioPathStatus
{
        optional bool connected = 1;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                  Select Sources Status and Source Slot Message Structure                     ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductSTSInterfaceStatus
{
        optional bool initialized = 1;
}

message ProductSelectSourceSlot
{
        optional ProductSTS.ProductSourceSlot slot = 1;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                               Network Status Message Structure                               ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductNetworkStatus
{
        enum ProductNetworkType
        {
             Wired    = 1;
             Wireless = 2;
             Unknown  = 3;
        }
        optional ProductNetworkType networktype = 1;
        optional bool               configured  = 2;
        optional bool               connected   = 3;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                           Network Wireless Status Message Structure                          ///
////////////////////////////////////////////////////////////////////////////////////////////////////
/// A separate message is created for sending the network wireless status. This message is       ///
/// populated with data from the Bose Network Manager process, and cannot be used to determine   ///
/// whether the Riviera board is connected to a network or not. Its primary purpose is to send   ///
/// data that determined whether a wireless network exists and is configurable, as well as to    ///
/// send frequency information to the LPM from the Product Controller to avoid any frequency     ///
/// interruption during a speaker Adapt IQ process.                                              ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductWirelessStatus
{
        optional bool   configured   = 1;
        optional uint32 frequencykhz = 2;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///              Voice Message Structure for the Virtual Personal Assistant VPA                  ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductVoiceStatus
{
        optional bool microphoneenabled = 1;
        optional bool accountconfigured = 2;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                                 Key Data Message Structure                                   ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductKeyData
{
        optional KeyActionPb.KEY_ACTION action = 1;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                             System Language Message Structure                                ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductLanguageData
{
        optional string systemlanguage = 1 [ default = "en" ];
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                           Configuration Status Message Structure                             ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductConfigurationStatus
{
        optional bool network  = 1 [ default = false ];
        optional bool language = 2 [ default = false ];
        optional bool account  = 3 [ default = false ];
}

message Language
{
    message Properties
    {
            repeated string supported_language_codes  = 1;
    }
    optional string     code       = 1;
    optional Properties properties = 2;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                              Autowake Status Message Structure                               ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductAutowakeStatus
{
        optional bool active = 1;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                              Playback Request Message Structure                              ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductPlaybackRequest
{
        enum ProductPlaybackState
        {
             Play    = 1;
             Pause   = 2;
             Stop    = 3;
             Unknown = 4;
        }
        optional ProductPlaybackState state = 1;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                                Base Product Message Structure                                ///
////////////////////////////////////////////////////////////////////////////////////////////////////
message ProductMessage
{
        optional ProductLpmStatus           lpmstatus           =  1;
        optional ProductCapsStatus          capsstatus          =  2;
        optional ProductAudioPathStatus     audiopathstatus     =  3;
        optional ProductSTSInterfaceStatus  stsinterfacestatus  =  4;
        optional ProductSelectSourceSlot    selectsourceslot    =  5;
        optional ProductNetworkStatus       networkstatus       =  6;
        optional ProductWirelessStatus      wirelessstatus      =  7;
        optional ProductVoiceStatus         voicestatus         =  8;
        optional ProductKeyData             keydata             =  9;
        optional ProductLanguageData        languagedata        = 10;
        optional ProductConfigurationStatus configurationstatus = 11;
        optional ProductAutowakeStatus      autowakestatus      = 12;
        optional ProductPlaybackRequest     playbackrequest     = 13;
        optional bool                       power               = 14;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
///                                        End of File                                           ///
////////////////////////////////////////////////////////////////////////////////////////////////////
